src/app/clusters/scenes-server/SceneHandlerImpl.h

```diff
@@ -0,0 +1,228 @@
+        {
+            uint16_t idx;
+            CHIP_ERROR err = FindPair(status.mEndpoint, idx);
+
+            if (CHIP_NO_ERROR == err)
+            {
+                mStatePairBuffer[idx] = status;
+            }
+            else if (mPairCount < MaxEndpointCount)
+            {
+                // If not found, insert at the end
+                mStatePairBuffer[mPairCount] = status;
+                mPairCount++;
+            }
+            else
+            {
+                return CHIP_ERROR_NO_MEMORY;
+            }
+
+            return CHIP_NO_ERROR;
+        }
+
+        CHIP_ERROR GetPair(const EndpointId endpoint, EndpointStatePair<ValueType> & status) const
+        {
+            uint16_t idx;
+            ReturnErrorOnFailure(FindPair(endpoint, idx));
+
+            status = mStatePairBuffer[idx];
+            return CHIP_NO_ERROR;
+        }
+
+        /// @brief Removes Pair and decrements Pair count if the endpoint existed in the array
+        /// @param endpoint : endpoint id of the pair
+        CHIP_ERROR RemovePair(const EndpointId endpoint)
+        {
+            uint16_t position;
+            VerifyOrReturnValue(CHIP_NO_ERROR == FindPair(endpoint, position), CHIP_NO_ERROR);
+
+            uint16_t nextPos = static_cast<uint16_t>(position + 1);
+            uint16_t moveNum = static_cast<uint16_t>(mPairCount - nextPos);
+
+            // Compress array after removal, if the removed position is not the last
+            if (moveNum)
+            {
+                memmove(&mStatePairBuffer[position], &mStatePairBuffer[nextPos], sizeof(EndpointStatePair<ValueType>) * moveNum);
+            }
+
+            mPairCount--;
+            // Clear the last occupied position
+            mStatePairBuffer[mPairCount].mEndpoint = kInvalidEndpointId;
+
+            return CHIP_NO_ERROR;
+        }
+
+        uint16_t mPairCount = 0;
+        EndpointStatePair<ValueType> mStatePairBuffer[MaxEndpointCount];
+    };
+
+    /// @brief Struct meant to allow to save context for a cluster that needs to apply a scene after a transition time but does not
+    /// own a time property in its commands
+    /// @tparam MaxEndpointCount
+    template <size_t MaxEndpointCount>
+    struct TransitionTimeInterface
+    {
+        EmberEventControl sceneHandlerEventControls[MaxEndpointCount];
+
+        TransitionTimeInterface(ClusterId clusterId, void (*callback)(EndpointId)) : mClusterId(clusterId), mCallback(callback) {}
+
+        /**
+         * @brief event control object for an endpoint
+         *
+         * @param[in] endpoint target endpoint
+         * @param[in] eventControlArray Array where to find the event control
+         * @return EmberEventControl* configured event control
+         */
+        EmberEventControl * getEventControl(EndpointId endpoint, const Span<EmberEventControl> & eventControlArray)
+        {
+            uint16_t index = emberAfGetClusterServerEndpointIndex(endpoint, mClusterId, MaxEndpointCount);
```

Comment:
This does not look right.  The third arg should be the fixed endpoint count, not the total endpoint count.

Reply:
And if tests did not catch this, we need better tests.

Reply:
Indeed, emberAfGetClusterServerEndpointIndex is not implemented for the unit tests. The yamls tests should catch it but I since the default CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT is 0, they currently wouldn't.

Enabling scenes on bridges or other devices with more than 0 dynamic endpoint will make the tests catch this though since we have a VeirfyOrDie in emberAfGetClusterServerEndpointIndex.